### DISCUSSION TITLE
ionospheric correction with stackSentinel.py

### DISCUSSION THREAD

**Discussion by SofiViotto** [isce-framework/isce2#579] [Q&A]
Hello everyone, 

I'm trying to create a new stack of interferograms and I would like to apply the ionospheric correction. The options are:
--param_ion ionosphere estimation parameter file. if provided, will do ionosphere estimation.
and
----num_connections_ion 

What file shall I provide to perfom the ionospheric correction?

I have tried the ionospheric correction before working with topsApp, but not with stackSentinel.py. Any advice?

Best,

Sofia

**Comment by AlejandroOroDiscord** [comment]
Hello Sofía! I performed the ionospheric correction with stackSentinel.py providing the **ion_param.txt** file located in /home/**YOURUSERNAME**/tools/src/isce2/contrib/stack/topsStack.
I am actually trying to do the opposite, I want to filter the ionosphere using topsApp.py. Could you share you topsApp.xml config file? cause I don't see much improvement with the parameters I use. Thanks in advance!

**Comment by SofiViotto** [comment]
Hello! Thanks for your reply. At the end, I figured out how to run it.
A question: what is the unit of the output ion_dates? Is it radian?

With topsapp I used everything by default. 

"do ionosphere correction" True
"start ionosphere step" subband
"end ionosphere step" esd 
"height of ionosphere layer in km" 200.0
 "apply polynomial fit before filtering ionosphere phase" True
  "maximum window size for filtering ionosphere phase" 200
  "minimum window size for filtering ionosphere phase" 100
"maximum window size for filtering ionosphere azimuth shift" 150
"minimum window size for filtering ionosphere azimuth shift" 75
"correct phase error caused by ionosphere azimuth shift" 1


I don't remember if the output interferogram, after the ionospheric correction, is corrected for the ionosphere or (as with the stripmap application) you have to subtract it afterwards. You may want to check that. Another reason why you don't see any improvement is that maybe you don't have a strong ionosphere impact. Normally for c-band the delay is so small that you can forget about it, but there are some cases (like my area of interest in NW Argentina and for ascending tracks) where you can really see some influence on the ionosphere. 


I hope it works for you!

Best,
Sofia

**Reply by AlejandroOroDiscord** [reply]
Thanks a lot for sharing your parameters Sofía! From what I read here:
https://github.com/isce-framework/isce2/blob/main/contrib/stack/topsStack/README.md
ion_dates must be in radians because it is phase. I hope someone can confirm this.
Also, thanks for your explanating about C-Band and the ionospheric effect. In my case study, San Juan, we have very high variations of height so I think it is very important to mitigate the ionospheric effect.

**Reply by SofiViotto** [reply]
I see. Are you working with ascending tracks? 

A quick tip: you can check the acquisition time of your data and the behaviour of the ionosphere. As the ionosphere often varies between seasons, but also during the day, you can expect a stronger influence on the acquisitions at certain times of the day (e.g. at sunrise).

**Reply by SofiViotto** [reply]
A question, going back to the ionosphere estimation with topsstack..  Have you seen any strange patterns in the resulting network? For example, I'm running everything by default (which means 3 connections). For some reason, a single date in the middle of the whole time span has only two connections.

**Reply by AlejandroOroDiscord** [reply]
I am using both orbits and right now I am focusing on the ascending orbit indeed. I will make sure to try out that tip you shared with me :).
In relation with the network, I used 7 connections and my network is also incomplete. For some acquisitions I have less connections than others. I don't know the reason but I know there are some requirements for the aquisitions to be used, for example, they are processed in groups that have the same starting and ending range:

From https://github.com/isce-framework/isce2/blob/main/contrib/stack/topsStack/README.md:
"In stack ionospheric phase estimation, acquistions with same swath starting ranges are put in a group. A network is formed within a group. Extra pairs are also processed to connect the different groups so that all acquistions are connected. But we need to estimate a phase offset for these extra pairs, which might not be accurate. Therefore, for a particualr swath starting ranges, if there are only a few acquistions, it's better to just discard them so that we don't have to estimate the phase offsets."

I used SNAP to look for acquisitions with different ranges and discarted them to make things simplier.

**Reply by SofiViotto** [reply]
Hi Alejandro! I just wanted to asked you a last thing.. What is the output resolution in terms of pixel size? To unwrap the interferogram 40 x10 (rg x az ) looks are applied. But then it aplies a new look of the ifg 5 x 5. Which means, at the end, 200 x 50 (Am I right?). Also, did you change it?

**Reply by AlejandroOroDiscord** [reply]
Hi Sofía! Ask me all the questions you need because I also need to ask a lot. For what I undestand, the first filter is multilooking which I set to 7 x 2 (rg x az). Then if you apply another filter of 5 x 5 to the same image, it should left you with a 35 x 10 (7 times 5 x 2 times 5) final reduction.
Hope this helps!
I would love to share some results with you so you can tell me what you think.

**Reply by AlejandroOroDiscord** [reply]
I am seeing that the results are the same with and without ionospheric correction estimated.
For topsApp.py I get this:
![topsApp py comparison lowRes](https://user-images.githubusercontent.com/100450317/193101437-8db4a277-e946-4965-9018-82cb4aec0dbd.jpg)
And for stackSentinel.py I get this for one of the filt_fine files:
![stackSentinel py comparison](https://user-images.githubusercontent.com/100450317/193102404-a4013a73-eba8-4a44-a967-0f1e810092d9.jpg)
Maybe I need to apply the correction to the results with an additional operation?

**Reply by SofiViotto** [reply]
Hey again...
Have you looked the ionosphere part only? 
I don't remember, to be honest, if topsApp does the correction or not. But I'm sure that for stripmap you would have to remove the ionospheric contribution at the end. And I'm also sure that the filt_fine.unw that you obtain with the stackSentinel App ARE NOT CORRECTED. For stackSentinel App you remove the contribution at the end.
You can do it with python.
For stackSentinel.py:
ion_dates/date1.ion - ion_dates/date2.ion = ion/date1_date2/ion_cal/filt.ion
ion_dates/date1.ion - ion_dates/date2.ion = ionospheric phase in merged/interferograms/date1_date2/filt_fine.unw

Best,
Sofi

**Reply by AlejandroOroDiscord** [reply]
Hi again Sofi! Thanks for your answer!!!! Then I am pretty sure I have to remove myself the contribution when using both, topsApp.py and stackSentinel.py.
So as I see it, for stackSentinel.py:
**filt.ion** is the ionospheric contribution.
**filt_fine.unw - filt.ion** is the unwrapped interferogram corrected by ionosphere.
Is this correct?
Also, is this how your filt.ion files looks like:
![20201128_20201210](https://user-images.githubusercontent.com/100450317/193401873-06aceedd-a28c-4cc3-9c31-56a8dddc5cab.png)

**Reply by fdelgadodelapuente** [reply]
> I am seeing that the results are the same with and without ionospheric correction estimated.
> For topsApp.py I get this:
> ![topsApp py comparison lowRes](https://user-images.githubusercontent.com/100450317/193101437-8db4a277-e946-4965-9018-82cb4aec0dbd.jpg)
> And for stackSentinel.py I get this for one of the filt_fine files:
> ![stackSentinel py comparison](https://user-images.githubusercontent.com/100450317/193102404-a4013a73-eba8-4a44-a967-0f1e810092d9.jpg)
> Maybe I need to apply the correction to the results with an additional operation?

That interferogram is not correctly processed because it includes the topographic phase. Check if your DEM is not empty.  You need to remove that before being able to observe any of the ionospheric signals.

**Reply by AlejandroOroDiscord** [reply]
Thanks for your answer Francisco! I tested different DEMs in SNAP. Copernicus 30m and SRTM 1Sec improved the results but none of them remove all the topographic contribution. I guess I should create my own DEM to get a better interferogram or is there any other option? These are my results:
SRTM3Sec:
![SRTM3Sec](https://user-images.githubusercontent.com/100450317/193965892-957e6c97-9515-48cf-b2bc-3d8b37fe0965.jpg)
SRTM1Sec:
![SRTM1Sec](https://user-images.githubusercontent.com/100450317/193966178-33d27fed-c2ee-48a5-bf5f-d758fb86fd64.jpg)
GETASSE30m:
![GETASSE30m](https://user-images.githubusercontent.com/100450317/193966285-be88e81a-7bb2-48c0-9518-9fb7dbaf93cf.jpg)
COP30m:
![COP30m](https://user-images.githubusercontent.com/100450317/193966390-5b9f4882-b262-41c0-979f-bda5027b0a2a.jpg)

**Reply by fdelgadodelapuente** [reply]
You should use the best available DEM, and those are SRTM 1 arcsec and Copernicus 1 arcsec. The interferograms processed with SRTM 3 arcsec and GETASSE30m have lots of high frequency signals that are topographic residuals. 

There is a ramp between the  SRTM 1 arcsec and Copernicus 1 arcsec  corrected interferograms and I suspect it could be related to missing corrections. The Copernicus DEM is distributed with the elevation with respect to the EGM2008 geoid and it has to be referenced with respect to the WGS84 spheroid in order to be compatible with ISCE.

**Reply by AlejandroOroDiscord** [reply]
Hello Francisco! You were right, the DEM was empty, the problem was the dem.py program not downloading the data. Now, with the SRTM1Sec I am getting the same results with ISCE2 than with SNAP:
![ISCE2_SRTM1Sec](https://user-images.githubusercontent.com/100450317/194323875-3ce664fc-0e69-4a9b-9f30-67ecd55f4d8f.jpg)
Thanks for the Copernicus DEM explanation! I think I need to find a better them to be able to completely remove the topographic phase contribution.

**Comment by fdelgadodelapuente** [comment]
Hi Alejandro,

That interferogram looks like an average pair for an area with high relief and lots of phase delays. Hence those signals are not topographic residuals.

Getting back to the point of this thread, the ionospheric signals for an ascending TOPS interferogram in areas of low geomagnetic latitude look like a weird ramp plus phase jumps between the different bursts after applying the ESD correction. This is shown clearly in Figure 5 of [Liang et al., IEEE, 2019 ](http://web.gps.caltech.edu/~simons/publications/pdfs/Liang_etal_2019.pdf). The interferogram that you show does not have those phase jumps. 

Francisco

**Reply by AlejandroOroDiscord** [reply]
Thank you Francisco for improving my interpretation of these interferograms! I read the paper you sent and I find it very interesting. I did notice weird fringing patterns but using more than one swath, in the examples I showed you above only subswath 2 of a Sentinel-1 image was processed and you don't see those features. I wonder if it has to do with using more than 1 sub-swath.
![WhatsApp Image 2022-09-24 at 11 03 11 marked](https://user-images.githubusercontent.com/100450317/194559610-ff3a920a-0c34-4e03-94b5-ed29bc5f5410.jpg)

**Reply by fdelgadodelapuente** [reply]
No, you will observe the ionospheric phase jumps in all the swaths that cover an area with a variation in the TEC. 

Another example of Ionospheric corrections for TOPS
https://ieeexplore.ieee.org/document/8114326

**Reply by AlejandroOroDiscord** [reply]
I see, thanks again for the information Francisco!

**Reply by birhan22** [reply]
where can I get TEC data?
